Some parser tests are being ignored

When running tests some of the parser tests are ignored due to a bug in CMake or CTest or the GoogleTest cmake module or something like that (https://gitlab.kitware.com/cmake/cmake/-/issues/21027): 
```
[0/1] Running tests...
Test project /github/workspace/build
CMake Error at /github/workspace/build/src/parser/test/sq-parser-test[1]_tests.cmake:129 (add_test):
  add_test called with incorrect number of arguments
Call Stack (most recent call first):
  /github/workspace/build/src/parser/test/sq-parser-test[1]_include.cmake:2 (include)
  /github/workspace/build/src/parser/test/CTestTestfile.cmake:7 (include)
  /github/workspace/build/src/CTestTestfile.cmake:10 (subdirs)
  CTestTestfile.cmake:8 (subdirs)


CMake Error at /github/workspace/build/src/parser/test/sq-parser-test[1]_tests.cmake:130 (set_tests_properties):
  set_tests_properties unknown error.
Call Stack (most recent call first):
  /github/workspace/build/src/parser/test/sq-parser-test[1]_include.cmake:2 (include)
  /github/workspace/build/src/parser/test/CTestTestfile.cmake:7 (include)
  /github/workspace/build/src/CTestTestfile.cmake:10 (subdirs)
  CTestTestfile.cmake:8 (subdirs)
```

--- a/src/parser/include/parser/Parser.h
+++ b/src/parser/include/parser/Parser.h
@@ -42,13 +42,13 @@ private:
   SQ_ND Ast *parse_dot_expression(Ast &parent);
   SQ_ND bool parse_field_call(Ast &parent);
   SQ_ND bool parse_parameter_pack(Ast &parent);
-  SQ_ND bool parse_parameter_list(Ast &parent);
-  SQ_ND bool parse_parameter(Ast &parent);
+  SQ_ND bool parse_parameter(Ast &parent, int &pos_count, int &named_count);
+  SQ_ND bool parse_positional_parameter(Ast &parent);
+  SQ_ND bool parse_named_parameter(Ast &parent);
   SQ_ND std::optional<Primitive> parse_primitive_value();
   SQ_ND std::optional<PrimitiveString> parse_dqstring();
   SQ_ND std::optional<PrimitiveBool> parse_bool();
   SQ_ND std::optional<PrimitiveFloat> parse_float();
-  SQ_ND bool parse_named_parameter(Ast &parent);
   SQ_ND bool parse_list_filter(Ast &parent);
   SQ_ND bool parse_slice_or_element_access(Ast &parent);
   SQ_ND bool parse_condition(Ast &parent);

--- a/src/parser/src/Parser.cpp
+++ b/src/parser/src/Parser.cpp
@@ -81,7 +81,7 @@ Ast *Parser::parse_dot_expression(Ast &parent) {
   return current_parent;
 }
 
-// field_call: Identifier parameter_list? list_filter?;
+// field_call: Identifier parameter_pack? list_filter?;
 bool Parser::parse_field_call(Ast &parent) {
   const auto opt_id = accept_token(Token::Kind::Identifier);
   if (!opt_id) {
@@ -93,37 +93,86 @@ bool Parser::parse_field_call(Ast &parent) {
   return true;
 }
 
-// parameter_pack: LParen parameter_list? Rparen
+// parameter_pack: LParen (
+//                    param_list_with_pos_params |
+//                    param_list_without_pos_params
+//                 )? RParen
+//
+// param_list_with_pos_params: positional_parameter
+//                             (Comma positional_parameter)*
+//                             (Comma named_parameter)*
+//
+// param_list_without_pos_params: named_parameter
+//                                (Comma named_parameter)*
+//
 bool Parser::parse_parameter_pack(Ast &parent) {
   if (!accept_token(Token::Kind::LParen)) {
     return false;
   }
-  (void)parse_parameter_list(parent);
+  // The code doesn't follow the same shape as the grammar (as written above)
+  // here because the grammar gets a bit messy but it's simple to say in code
+  // or English: A list of parameters, separated by commas, where parameters
+  // can be either positional or named, and all of the positional parameters
+  // must come before all of the named parameters.
+  auto pos_count = 0;
+  auto named_count = 0;
+  while (parse_parameter(parent, pos_count, named_count)) {
+  }
   (void)expect_token(Token::Kind::RParen);
   return true;
 }
 
-// parameter_list: parameter (Comma parameter)*
-bool Parser::parse_parameter_list(Ast &parent) {
-  if (!parse_parameter(parent)) {
+// Doesn't represent a symbol in the grammar because it's easier to code this
+// way.
+// Parses the next parameter in a parameter list, whether it's the first
+// argument or not, and whether it's positional or named.
+bool Parser::parse_parameter(Ast &parent, int &pos_count, int &named_count) {
+  // Require a comma unless we're parsing the first parameter
+  if ((pos_count + named_count) > 0 && !accept_token(Token::Kind::Comma)) {
     return false;
   }
-  while (accept_token(Token::Kind::Comma)) {
-    if (!parse_parameter(parent)) {
-      throw ParseError(tokens_.read(), expecting_);
-    }
+  // Only parse a positional parameter if we haven't seen any named ones yet.
+  if (named_count == 0 && parse_positional_parameter(parent)) {
+    ++pos_count;
+    return true;
   }
-  return true;
+  if (parse_named_parameter(parent)) {
+    ++named_count;
+    return true;
+  }
+  if ((pos_count + named_count) > 0) {
+    // We've seen a comma without a parameter after it
+    throw ParseError(tokens_.read(), expecting_);
+  }
+  return false;
 }
 
-// parameter: (primitive_value | named_parameter)
-bool Parser::parse_parameter(Ast &parent) {
+// positional_parameter: primitive_value
+bool Parser::parse_positional_parameter(Ast &parent) {
   if (auto opt_prim = parse_primitive_value()) {
     parent.data().params().pos_params().emplace_back(
         std::move(opt_prim.value()));
     return true;
   }
-  return parse_named_parameter(parent);
+  return false;
+}
+
+// named_parameter: Identifier Equals primitive_value
+bool Parser::parse_named_parameter(Ast &parent) {
+  const auto opt_id_token = accept_token(Token::Kind::Identifier);
+  if (!opt_id_token) {
+    return false;
+  }
+  const auto &id = opt_id_token.value().view();
+
+  (void)expect_token(Token::Kind::Equals);
+
+  auto opt_prim = parse_primitive_value();
+  if (!opt_prim) {
+    throw ParseError(tokens_.read(), expecting_);
+  }
+  parent.data().params().named_params().emplace(id, opt_prim.value());
+  return true;
 }
 
 // primitive_value: (Integer | DQString | bool)
@@ -197,24 +246,6 @@ std::optional<PrimitiveFloat> Parser::parse_float() {
   // has already told us that the token matches the pattern for a float.
   ASSERT(str_end == float_cstr + float_str.size());
   return ret;
-}
-
-// named_parameter: Identifier Equals primitive_value
-bool Parser::parse_named_parameter(Ast &parent) {
-  const auto opt_id_token = accept_token(Token::Kind::Identifier);
-  if (!opt_id_token) {
-    return false;
-  }
-  const auto &id = opt_id_token.value().view();
-
-  (void)expect_token(Token::Kind::Equals);
-
-  auto opt_prim = parse_primitive_value();
-  if (!opt_prim) {
-    throw ParseError(tokens_.read(), expecting_);
-  }
-  parent.data().params().named_params().emplace(id, opt_prim.value());
-  return true;
 }
 
 // list_filter: LBracket (slice_or_element_access | condition) RBracket

--- a/src/parser/test/test_parser.cpp
+++ b/src/parser/test/test_parser.cpp
@@ -12,6 +12,7 @@
 #include "util/strutil.h"
 
 #include <gtest/gtest.h>
+#include <iostream>
 #include <limits>
 #include <string_view>
 #include <utility>
@@ -107,8 +108,18 @@ TEST(AstTest, MultipleEntrypoints) {
   expect_plain_leaf(b, "b");
 }
 
-using SimpleTestCase =
-    std::tuple<std::string_view, FieldCallParams, FilterSpec>;
+struct SimpleTestCase {
+  std::string_view query_;
+  FieldCallParams fcp_;
+  FilterSpec filter_spec_;
+};
+
+std::ostream &operator<<(std::ostream &os, const SimpleTestCase &tc) {
+  os << "(query=" << tc.query_ << ";fcp=" << tc.fcp_
+     << ";filter=" << util::variant_to_str(tc.filter_spec_) << ")";
+  return os;
+}
+
 class SimpleAstTest : public testing::TestWithParam<SimpleTestCase> {};
 
 TEST_P(SimpleAstTest, SimpleTest) {
@@ -417,7 +428,7 @@ INSTANTIATE_TEST_SUITE_P(
         SimpleTestCase{"a(false_n=1)", params(named("false_n", 1)),
                        no_filter_spec}));
 
-class InvalidQueryTest : public testing::TestWithParam<std::string_view> {};
+class InvalidQueryTest : public testing::TestWithParam<const char *> {};
 
 TEST_P(InvalidQueryTest, InvalidQuery) {
   try {
@@ -427,14 +438,13 @@ TEST_P(InvalidQueryTest, InvalidQuery) {
   }
 }
 
-INSTANTIATE_TEST_SUITE_P(
-    InvalidQueryTestInstantiation, InvalidQueryTest,
-    testing::Values(".a", "a.", "{", "a {", "}", "a }", "a { }", "{ }", "{ } a",
-                    "{ } a", "a(", "a)", "a(1", "a(\"str)", "a(p=)", "a(p)",
-                    "a(p=1,2)"
-                    "a(p-x)",
-                    "a[", "a]", "a[]", "a[=]", "a[10=]", "a[1.0]", "a[1.0:]",
-                    "a[:1.0:]", "a[::1.0]"));
+INSTANTIATE_TEST_SUITE_P(InvalidQueryTestInstantiation, InvalidQueryTest,
+                         testing::Values(".a", "a.", "{", "a {", "}", "a }",
+                                         "a { }", "{ }", "{ } a", "{ } a", "a(",
+                                         "a)", "a(1", "a(\"str)", "a(p=)",
+                                         "a(p)", "a(p=1,2)", "a(p-x)", "a[]",
+                                         "a[=]", "a[10=]", "a[1.0]", "a[1.0:]",
+                                         "a[:1.0:]", "a[::1.0]", "a(\")"));
 
 } // namespace
 } // namespace sq::test

--- a/src/results/test/test_results.cpp
+++ b/src/results/test/test_results.cpp
@@ -17,6 +17,7 @@
 
 #include <gsl/gsl>
 #include <gtest/gtest.h>
+#include <iostream>
 #include <range/v3/view/cartesian_product.hpp>
 #include <range/v3/view/iota.hpp>
 #include <range/v3/view/transform.hpp>
@@ -264,7 +265,16 @@ TEST_F(ResultTreeTest, TestGeneratedTreeWithMultipleCallsPerObject) {
 // Param passing tests
 // -----------------------------------------------------------------------------
 
-using ParamPassingTestCase = std::tuple<std::string_view, FieldCallParams>;
+struct ParamPassingTestCase {
+  const char *query_;
+  FieldCallParams fcp_;
+};
+
+std::ostream &operator<<(std::ostream &os, const ParamPassingTestCase &pptc) {
+  os << "(query=" << pptc.query_ << ",fcp=" << pptc.fcp_ << ")";
+  return os;
+}
+
 struct ResultTreeParamTest
     : MockFieldTest,
       ::testing::WithParamInterface<ParamPassingTestCase> {};

